stablehlo/dialect/TypeInference.cpp

```diff
@@ -1551,9 +1506,155 @@ LogicalResult inferTriangularSolveOp(
   return success();
 }
 
-LogicalResult inferWhileOp(Optional<Location> location, ValueRange operand,
-                           Region& cond, Region& body,
+LogicalResult inferWhileOp(Optional<Location>, ValueRange operand,
                            SmallVectorImpl<Type>& inferredReturnTypes) {
+  for (const auto& resultType : operand.getType())
+    inferredReturnTypes.push_back(resultType);
+  return success();
+}
+
+// We intend to verify the following properties
+//  P1. Verify all `inputs` need to have compatible shapes.
+//  P2. Verify that
+//      1. the dimensions of reduce-op are in-bounds for the given shape.
+//      2. the dimension-attribute have no duplicate entries.
+//  P3. Verify the inner block defining the reducer function.
+LogicalResult verifyReduceOp(Optional<Location> location, ValueRange inputs,
+                             ValueRange initValues,
+                             DenseIntElementsAttr dimensions, Region& body) {
+  SmallVector<TensorType> inputArgTypes{llvm::map_range(
+      inputs.getTypes(),
+      [](Type t) -> TensorType { return t.cast<TensorType>(); })};
+  SmallVector<TensorType> initValueTypes{llvm::map_range(
+      initValues.getTypes(),
+      [](Type t) -> TensorType { return t.cast<TensorType>(); })};
+
+  // P1. & P2.
+  SmallVector<int64_t> newDimensions;
+  if (failed(verifyReduceOpInputsAndInferShape(
```

Comment:
Let me defense with these:
1. As far as digging to code, `verify()` runs before `inferReturnTypeComponents()`, I have shared you with a doc of my founds.
2. This PR aims to address another tricky case that:  the shape function is called **alone** without the verifier: thus we have no choice but to duplicate these verification logic. See https://github.com/openxla/stablehlo/issues/623 

Thanks for pointing out the final part in the comment! Could you open a bug for it and elaborate a bit? I feel it's an optimization, and could do it when we have more cycles. Please correct me otherwise.

Reply:
> 1. As far as digging to code, `verify()` runs before `inferReturnTypeComponents()`, I have shared you with a doc of my founds.
you are right. I didn't realize that InferTypeOpInterface sets verifyWithRegions to `1`. 

That means that we need to move all the verification to verifyRegions by using `hasRegionVerifier`. By the time it reaches verifyRegions, inference has been verified so we don't have to do those checks again.